### DISCUSSION TITLE
docker-compose: compatibility issues, missing features, and other thoughts

### DISCUSSION THREAD

**Discussion by gotmax23** [containers/podman#10976] [General]
I am glad that Podman now supports docker-compose for both rootful and rootless containers! However, there are a couple missing features and compatibility issues that make it difficult for me to use docker-compose to manage Podman containers. I'm not sure what the process is for resolving these issues upstream. Is it even possible to implement Podman specific features, given that docker-compose uses the **Docker** Python api? Also, I'm not sure how willing the docker-compose developers are to cater to Podman.

It would be awesome if we could fix this! Let me know if I should file a formal bug report. In the meantime, here are the issues I found:

1. The `--systemd=true|false|always` option is not implemented.
  
2. It is not possible to specify `--userns` options, such as `--userns keep-id`. Running containers with `--userns keep=-id` and `--user "$(id -u):$(id -g)"` is the best way I've found to allow the user who creates a rootless container to access files in a bind mount without `podman unshare` or setting `--user=root`.
  
3. docker-compose cannot build images using Podman. Please see [this issue](https://github.com/docker/compose/issues/8304) for more info.
  
4. Containers managed through docker-compose do not restart on boot. I assume the recommended solution is to create a Systemd unit that runs `docker-compose up` at startup. This is a clunky solution, in my opinion. It would be better if restarting containers on boot was handled automatically. Would it be possible to implement a seperate Systemd service that starts containers at boot, as @rhatdan suggested in [this issue](https://github.com/containers/podman/issues/6508#issuecomment-640085744)? The `containers.podman.podman_container` Ansible module is not able to create containers that start at boot, either. ([It does not support creating systemd units](https://github.com/containers/ansible-podman-collections/issues/18).) Implementing this feature would solve both problems.
  
5. Docker will soon release ["Compose v2"](https://github.com/docker/compose-cli) which is written in Go and has a different architecture than the current iteration. The plan is for this to replace the current python-based docker-compose project. How will this effect Podman?
  

Thanks,
Maxwell

**Comment by mheon** [comment]
For your first two questions - what exactly is the ask here? Implement these options in `docker-compose` itself? I doubt we'll be able to do so, given they primarily target the Docker engine and it does not support either of them, and we can't easily extend the Docker API that compose talks to given that we don't control its versioning. I don't know that we can easily offer an out-of-band way of setting these, either, outside of setting the default for the flags via `containers.conf`.

For your third question - would the `podman-docker` wrapper be enough to solve that? Again, it's not something the Podman project can really solve ourselves, but here we can offer the option to transparently replace the Docker binary.

For your fourth question, please see https://github.com/containers/podman/issues/10539.

For your fifth - we're still evaluating right now. Last I checked, Compose v2 was still experimental, so we're not terribly concerned yet. The project being in Go does offer some potentially exciting options (vendoring the Compose code into Podman itself to offer a `podman compose` command is something I've tossed around, for example, though I haven't actually looked into what would be required to make this feasible). In short, we have no updates as of yet, but we're aware of Compose v2 and are starting to look at it.

**Reply by gotmax23** [reply]
Thank you for your helpful response!

>  I don't know that we can easily offer an out-of-band way of setting these, either, outside of setting the default for the flags via containers.conf.

How would I do this for the `--userns` and `--systemd` flags? I looked at `man containers.conf` and the default at `/usr/share/containers/containers.conf` and did not see this option.

**Reply by mheon** [reply]
I do see a `userns` option in containers.conf which should correspond to `--userns` but unfortunately there is no `--systemd ` option as yet that I can see.

**Reply by rhatdan** [reply]
Open an issue in podman for the systemd option.
